### DISCUSSION TITLE
Wild delta BT swings and slow reading

### DISCUSSION THREAD

**Discussion by blondica73** [artisan-roaster-scope/artisan#1194] [Q&A] [answered]
I've been roasting for awhile; however, I wanted to automate my roaster and noticed something that puzzled me. I turned off smoothing so that I get the raw error and it appears that delta BT is "sticky", it swings between 0 and large values. If I enable smoothing and delta spans it seems to quiet down.

![Automation first attempt](https://github.com/artisan-roaster-scope/artisan/assets/40738317/55f41ab0-405a-4e1a-bca5-0474c399f103)

here were my settings ( i did play with smoothing during roast)
![Automation first Curves](https://github.com/artisan-roaster-scope/artisan/assets/40738317/654abbd7-0609-4ddd-914d-a426c4161d99)

I checked system noise and looks reasonable
![noise](https://github.com/artisan-roaster-scope/artisan/assets/40738317/1870e153-7d1b-4ab3-811f-8f856f3a110e)

**Comment by bigl00z3** [comment]
Detla BT or RoR is only the derivative of your BT signal. you can clearly see the steps on your BT curves. However it is not all the time through yrou recording. It looks like they are times it is recording fine. During the bumpy parts, looking at the values it could be:

- Your data logger only returns less precise data (e.g. without any decimals) due to unexplained conditions
- Your computer is overloaded / slowing down and the sample rate is reduced with only 1 data point every 2-3 sec

Questions you can ask yourself:

- Did you turn anything on or off that could electrically disturb the signal (e.g. turn a motor on, connect the power supply on your laptop...)
- Is your computer or the datalogger too slow at some points and cannot return or read enough values? 

Anyway if you want others to be able to help you you should:

- Give more details about your setup (Roaster type, what type of equipment isconnected to Artisan).
- Upload your artisan-settings to your post for others to analyse. Export them via `Help/Save Settings...` and rename the artisan-settings.aset  with a .txt extension before uploading on the GitHub discussion page. Don't send it via email.
- 
Nico

**Reply by blondica73** [reply]
I currently use a Windows system with a Dell laptop that has four cores, a very fast laptop.  I connect to the roaster, a BC-5 roaster, via Bluetooth.  I did notice that the temperature updates slowly in Artisan. I just don't know if that is a roaster communication issue, or something else. I run my laptop on battery, and I even connected a USB cable to the Buckeye-5 roaster, with no luck. 

Per your request, I attached the artisan-settings file
[artisan-settings.txt](https://github.com/artisan-roaster-scope/artisan/files/11689544/artisan-settings.txt)

**Reply by bigl00z3** [reply]
ok, I looked at your settings. You have a sampling rate of 1 second which is fast. On a fast computer, it should not create an issue. 

I don't understand why your BT signal suddenly has those steps. Could it be that the Modbus data is suddenly less precise (e.g. from 1 decimal to no decimal on the temperature value)? You need to find a way not to have those steps happening. and looking at it I doubt it is Artisan (others correct me if you thing differently) but most likely the roaster not communicating enough data.

Anyway, having a delta span (the duration over which the RoR is calculated) of zero will on most roaster give you a lot of noise on the signal, resulting in pretty much unusable curves. Try calculating it over 10 or 15seconds. Even without smoothing your signal will be much better. 

On my roaster I find that a Delta Span of 15 seconds pretty good.

Nico

**Reply by bigl00z3** [reply]
additional thought, you could check if changing the modbus settings in the `Config/Port...`  helps with the recorded data.
Try ticking the `fetch full blocks` to see if it helps:
![image](https://github.com/artisan-roaster-scope/artisan/assets/8886922/c8ff61e8-3177-451f-a464-19b8ee0656a6)

**Reply by blondica73** [reply]
I just tried that, and it does not make a difference. The temperature readings on my Artisan screen only update once every 3 seconds.

**Reply by poundy** [reply]
you said you've been roasting for a while. Does that mean you've been using Artisan for a while? Or are you only now implementing Artisan to read from the roaster? 

I have a BK 2.5kg roaster, nominally the same as the BC5. The machine setup from https://github.com/artisan-roaster-scope/artisan/blob/master/src/includes/Machines/BC/Roaster.aset is setting a sampling interval of 2sec (not 1) and no oversampling. I believe you're overloading the roaster PID serial connection with data requests and that's causing you to miss samples. Thats why I wanted to know about the history - my assumption is that this isn't a new phenomenon, it's been evident since you started trying to log this data.

**Comment by CarefreeBuzzBuzz** [comment]
I recall that the Bluetooth  connections readings are slow.  Can you try connecting in a hardwired manner to see if that is it?___________________________Michael Herbert, CarefreeBuzzBuzzOn Jun 8, 2023, at 8:01 AM, blondica73 ***@***.***> wrote:﻿
I just tried that, and it does not make a difference. The temperature readings on my Artisan screen only update once every 3 seconds.

**Reply by blondica73** [reply]
I bought a couple of RTD probes and the Phidgets hub and I can now sample at 0.5s. This leads me to believe that the communication hub is just slow on my roaster.

**Reply by poundy** [reply]
Cool, it's ultimately your money to waste. Sampling this fast isn't going to help you make better coffee :)

**Reply by blondica73** [reply]
It will allow me to test my automation set up.

**Reply by blondica73** [reply]
I would still want to debug this. Why can I read the Phidgets set up at 0.5s sampling rate, but cannot do that on the stock USB, even with cable attached, no difference?

**Reply by MAKOMO** [reply]
Isn't the hardware (processor), the software/firmware and the protocol of the Phidgets totally different from what you have on the other side of the stock USB? MODBUS RTU hardly allows sampling at 1sec with any device. One part is that this is plain old serial communication over USB vs native USB communication as for the case of Phidgets.

**Reply by poundy** [reply]
Agree fully with @MAKOMO here, this is two completely different hardware implementations that do completely different things. (plus, sampling so fast doesn't make things "better")

**Reply by blondica73** [reply]
Thank you for the explanation. Therefore, you believe that the MODUS RTU on my machine cannot be sampled faster than 1-2 seconds? So this is a hardware issue?

**Reply by poundy** [reply]
"issue"? no. Limitation, yes.

**Reply by blondica73** [reply]
Yes, limitations.

**Reply by MAKOMO** [reply, marked as answer]
Moreover, sampling temperature elements at 0.5sec is close to the limit what temperature hardware (the chip reading the probe) takes to gather readings (without any communication) and most hardware prefers to sample internally several samples which are forwarded as average to reduce noise. This implies that 1sec sampling interval for roasting using non-aerospace sensors is a useful limit. Artisan allows to sample faster for applications beyond roasting.

**Comment by blondica73** [comment]
> Isn't the hardware (processor), the software/firmware and the protocol of the Phidgets totally different from what you have on the other side of the stock USB? MODBUS RTU hardly allows sampling at 1sec with any device. One part is that this is plain old serial communication over USB vs native USB communication as for the case of Phidgets.



> "issue"? no. Limitation, yes.



> Moreover, sampling temperature elements at 0.5sec is close to the limit what temperature hardware (the chip reading the probe) takes to gather readings (without any communication) and most hardware prefers to sample internally several samples which are forwarded as average to reduce noise. This implies that 1sec sampling interval for roasting using non-aerospace sensors is a useful limit. Artisan allows to sample faster for applications beyond roasting.
